hazelcast/src/main/java/com/hazelcast/instance/impl/Node.java

```diff
@@ -765,23 +805,86 @@ public class NodeShutdownHookThread extends Thread {
+                        hazelcastInstance.getLifecycleService().terminate();
+                        break;
+                    case GRACEFUL:
+                        hazelcastInstance.getLifecycleService().shutdown();
+                        break;
+                    default:
+                        throw new IllegalArgumentException("Unimplemented shutdown hook policy: " + policy);
                 }
             } catch (Exception e) {
                 logger.warning(e);
             }
         }
+
+        private void shutdownWithIntent(ClusterTopologyIntent shutdownIntent) {
+            // consider the detected shutdown intent before triggering node shutdown
+            if (shutdownIntent == ClusterTopologyIntent.STABLE
+                || shutdownIntent == MISSING_MEMBERS) {
+                try {
+                    long timeoutNanos = getProperties().getNanos(ClusterProperty.CLUSTER_SHUTDOWN_TIMEOUT_SECONDS);
+                    while (!nodeEngine.getPartitionService().isPartitionTableHealthy()
+                            && timeoutNanos > 0) {
+                        sleep(TimeUnit.SECONDS.toMillis(1));
+                        timeoutNanos -= TimeUnit.SECONDS.toNanos(1);
+                    }
+                    changeClusterState(clusterStateForMissingMembers);
+                } catch (Throwable t) {
+                    // let shutdown proceed even though we failed to switch to FROZEN state
+                    logger.warning("Could not switch to transient FROZEN state while cluster"
+                            + "shutdown intent was " + shutdownIntent, t);
+                }
+            } else if (shutdownIntent == CLUSTER_SHUTDOWN
+                    || shutdownIntent == CLUSTER_SHUTDOWN_WITH_MISSING_MEMBERS) {
+                try {
+                    changeClusterState(ClusterState.PASSIVE);
+                } catch (Throwable t) {
+                    // let shutdown proceed even though we failed to switch to PASSIVE state
+                    // and wait for replica sync
+                    logger.warning("Could not switch to transient PASSIVE state while cluster"
+                            + "shutdown intent was " + shutdownIntent, t);
+                }
+                long timeoutNanos = getProperties().getNanos(ClusterProperty.CLUSTER_SHUTDOWN_TIMEOUT_SECONDS);
+                try {
+                    getNodeExtension().getInternalHotRestartService()
+                            .waitPartitionReplicaSyncOnCluster(timeoutNanos, TimeUnit.NANOSECONDS);
+                } catch (IllegalStateException e) {
+                    logger.severe("Failure while waiting for partition replica sync before shutdown", e);
+                }
+            }
+        }
+    }
+
+    public ClusterTopologyState getClusterTopologyState() {
+        return clusterTopologyState.get();
+    }
+
+    public void initializeClusterTopologyIntent(ClusterTopologyState clusterTopologyState) {
```

Comment:
seems this method is not used, is this intentional?

Reply:
it is used on EE side (https://github.com/hazelcast/hazelcast-enterprise/pull/5140/files#diff-2da293d6c9f59632497f3c8cabe2ec3bd0cbce9baf430a573ba108dc110f7116R1336)